controllers/opentelemetrycollector_controller.go

```diff
@@ -114,6 +123,124 @@ func (r *OpenTelemetryCollectorReconciler) removeRouteTask(ora autodetect.OpenSh
 	return nil
 }
 
+func (r *OpenTelemetryCollectorReconciler) doCRUD(ctx context.Context, params manifests.Params) error {
+	// Collect all objects owned by the operator, to be able to prune objects
+	// which exist in the cluster but are not managed by the operator anymore.
+	pruneObjects, err := r.findObjectsOwnedByOtelOperator(ctx, params.Instance)
+	if err != nil {
+		return err
+	}
+	managedObjects, err := r.BuildAll(params)
+	if err != nil {
+		return err
+	}
+	var errs []error
+	for _, obj := range managedObjects {
+		l := r.log.WithValues(
+			"object_name", obj.GetName(),
+			"object_kind", obj.GetObjectKind(),
+		)
+
+		if isNamespaceScoped(obj) {
+			if setErr := ctrl.SetControllerReference(&params.Instance, obj, params.Scheme); setErr != nil {
+				l.Error(setErr, "failed to set controller owner reference to resource")
+				errs = append(errs, setErr)
+				continue
+			}
+		}
+
+		desired := obj.DeepCopyObject().(client.Object)
+		mutateFn := manifests.MutateFuncFor(obj, desired)
```

Comment:
After rereading this, I think the naming is confusing. `managedObjects` is really the *desired* objects, and inside the loop we copy the loop variable and name that *desired*. The values we pass to `MutateFuncFor` are identical anyway, so this ends up not mattering, but it makes understanding this code unnecessarily difficult imo. How about:

```suggestion
	desiredObjects, err := r.BuildAll(params)
	if err != nil {
		return err
	}
	var errs []error
	for _, desired := range desiredObjects {
		l := r.log.WithValues(
			"object_name", obj.GetName(),
			"object_kind", obj.GetObjectKind(),
		)

		if isNamespaceScoped(obj) {
			if setErr := ctrl.SetControllerReference(&params.Instance, obj, params.Scheme); setErr != nil {
				l.Error(setErr, "failed to set controller owner reference to resource")
				errs = append(errs, setErr)
				continue
			}
		}
                
                // existing is an object the controller runtime will hydrate for us
                // we obtain it by deep copying the desired object because it's the most convenient way
		existing := obj.DeepCopyObject().(client.Object)
		mutateFn := manifests.MutateFuncFor(existing, desired)
```

Reply:
yeah i like that! adding...

Reply:
so i had to change this a bit, because the mutate you had actually reversed the parameters. What i did was just renamed desiredObjects to expectedObjects which is more accurate.